src/lib/charm/openstack/barbican.py

```diff
@@ -91,68 +169,141 @@ class BarbicanAdapters(charms_openstack.adapters.OpenStackRelationAdapters):
     This plumbs in the BarbicanConfigurationAdapter as the ConfigurationAdapter
     to provide additional properties.
     """
+
+    relation_adapters = {
+        'hsm': HSMAdapter,
+    }
+
     def __init__(self, relations):
         super(BarbicanAdapters, self).__init__(
-            relations, options=BarbicanConfigurationAdapter)
+            relations, options_instance=BarbicanConfigurationAdapter())
 
 
 class BarbicanCharm(charms_openstack.charm.OpenStackCharm):
     """BarbicanCharm provides the specialisation of the OpenStackCharm
     functionality to manage a barbican unit.
     """
 
-    release = 'liberty'
+    release = 'mitaka'
     name = 'barbican'
     packages = PACKAGES
     api_ports = {
-        'barbican-api': {
+        'barbican-worker': {
             os_ip.PUBLIC: 9311,
             os_ip.ADMIN: 9312,
-            os_ip.INTERNAL: 9313,
+            os_ip.INTERNAL: 9311,
         }
     }
     service_type = 'barbican'
-    default_service = 'barbican-api'
-    services = ['barbican-api', 'barbican-worker']
+    default_service = 'barbican-worker'
+    services = ['apache2', 'barbican-worker']
 
+    # Note that the hsm interface is optional - defined in config.yaml
     required_relations = ['shared-db', 'amqp', 'identity-service']
 
     restart_map = {
-        "{}/{}".format(BARBICAN_DIR, BARBICAN_API_CONF): services,
-        "{}/{}".format(BARBICAN_DIR, BARBICAN_ADMIN_PASTE_CONF): services,
-        "{}/{}".format(BARBICAN_DIR, BARBICAN_API_PASTE_CONF): services,
+        BARBICAN_CONF: services,
+        BARBICAN_API_PASTE_CONF: services,
     }
 
     adapters_class = BarbicanAdapters
 
-    def __init__(self, release=None, **kwargs):
-        """Custom initialiser for class
-
-        If no release is passed, then the charm determines the release from the
-        ch_utils.os_release() function.
-
-        Note that the release_selector can be used to control which class is
-        instantiated.
-        """
-        if release is None:
-            # release = ch_utils.os_release('barbican-common')
-            release = ch_utils.os_release('python-keystonemiddleware')
-        super(BarbicanCharm, self).__init__(release=release, **kwargs)
-
     def install(self):
         """Customise the installation, configure the source and then call the
         parent install() method to install the packages
         """
-        charmhelpers.fetch.add_source("ppa:gnuoy/barbican-alt")
+        # DEBUG - until seed random change lands into xenial cloud archive
+        # BUG #1599550 - barbican + softhsm2 + libssl1.0.0:
+        #  pkcs11:_generate_random() fails
+        charmhelpers.fetch.add_source("ppa:ajkavanagh/barbican")
```

Comment:
We can't release the charm into stable with this present. Maybe add a warning here?